test/unit/gatsby-plugin-image-hook.test.ts

```diff
@@ -351,6 +352,282 @@ describe('gatsby-plugin-image hook', () => {
       expect(actual.images?.fallback?.sizes).toMatch('100vw');
     });
   });
+
+  describe(`breakpoints generation`, () => {
+    describe(`layout: 'fixed'`, () => {
+      test(`should generate widths at 1x, 2x, 3x, 3x, and 5x width`, () => {
+        const actual = generateBreakpoints({
+          layout: 'fixed',
+          width: 100,
+        });
+
+        expect(actual.breakpoints).toEqual([100, 200, 300, 400]);
+      });
+      test(`should reduce quality when increasing size`, () => {
+        const actual = generateBreakpoints({
+          layout: 'fixed',
+          width: 100,
+        });
+
+        if (actual.breakpointsWithData == null) {
+          fail('actual.breakpointsWithData not set');
+          return;
+        }
+        expect(actual.breakpointsWithData.length).toBeGreaterThan(1);
+        actual.breakpointsWithData?.map((v, i) => {
+          if (i === 0) {
+            return;
+          }
+
+          expect(v.quality).toBeLessThan(
+            (actual?.breakpointsWithData || [])[i - 1].quality,
+          );
+        });
+      });
+      test(`should be able to disable variable quality`, () => {
+        const actual = generateBreakpoints({
+          layout: 'fixed',
+          width: 100,
+          disableVariableQuality: true,
+        });
+
+        expect(
+          actual.breakpointsWithData == null ||
+            actual.breakpointsWithData.length === 0,
+        ).toBe(true);
+      });
+      test(`should not generate widths larger than sourceWidth`, () => {
+        const actual = generateBreakpoints({
+          layout: 'fixed',
+          width: 100,
+          sourceWidth: 250,
+        });
+
+        actual.breakpoints.map((width) => {
+          expect(width).toBeLessThanOrEqual(250);
+        });
+        actual?.breakpointsWithData?.map(({ width }) => {
+          expect(width).toBeLessThanOrEqual(250);
+        });
+      });
+      test(`should not generate widths larger than 8192px, even if source width is larger`, () => {
+        const actual = generateBreakpoints({
+          layout: 'fixed',
+          width: 5000,
+          sourceWidth: 15000,
+        });
+
+        actual.breakpoints.map((width) => {
+          expect(width).toBeLessThanOrEqual(8192);
+        });
+        actual?.breakpointsWithData?.map(({ width }) => {
+          expect(width).toBeLessThanOrEqual(8192);
+        });
+      });
+    });
+    describe(`layout: 'constrained'`, () => {
+      test(`should generate widths that grow at ~8%`, () => {
+        const actual = generateBreakpoints({
+          layout: 'constrained',
+          width: 300,
+        });
+
+        expect(actual.breakpoints).toEqual([
```

Comment:
Would it make sense to ensure that the list always includes `width` multipled all the dpr values, so that there's always a source that doesn't need resizing in the browser.

Reply:
Yeah I can get onboard with this suggestion